README
NAME
    DateTime::Format::EMIUCP - Parse time formats for EMI-UCP protocol

SYNOPSIS
      use DateTime::Format::EMIUCP;

      my $scts = DateTime::Format::EMIUCP->parse_datetime('030212065530');
      print $scts->ymd; # 2012-02-03
      print $scts->hms; # 06:55:30

      my $vp = DateTime::Format::EMIUCP->parse_datetime('0302120655');
      print $vp->ymd; # 2012-02-03
      print $vp->hms; # 06:55:00

DESCRIPTION
    These formats are part of EMI-UCP protocol message. EMI-UCP protocol is
    primarily used to connect to short message service centers (SMSCs) for
    mobile telephones.

    SCTS is a string of 12 numeric characters which represents Service
    Center time-stamp in ddMMyyHHmmss format.

    DDT is a string of 10 numeric characters which represents deferred
    delivery time in ddMMyyHHmm format.

    VP is a string of 10 numeric characters which represents validity period
    time in ddMMyyHHmm format.

    See EMI-UCP Interface 5.2 Specification for further explanations.

PREREQUISITES
    * DateTime::Format::Builder

SEE ALSO
    DateTime::Format::EMIUCP::DDT, DateTime::Format::EMIUCP::SCTS,
    DateTime::Format::EMIUCP::VP, DateTime.

BUGS
    If you find the bug or want to implement new features, please report it
    at http://rt.cpan.org/NoAuth/Bugs.html?Dist=DateTime-Format-EMIUCP

    The code repository is available at
    http://github.com/dex4er/perl-DateTime-Format-EMIUCP

AUTHOR
    Piotr Roszatycki <dexter@cpan.org>

LICENSE
    Copyright (c) 2012 Piotr Roszatycki <dexter@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as perl itself.

    See http://dev.perl.org/licenses/artistic.html



Changes
------------------------------------------------------------------------
0.0200 | Piotr Roszatycki <dexter@cpan.org> | 2012-04-15

* Separate filters for SCTS, DDT and VP fields for EMI-UCP protocol.

------------------------------------------------------------------------
0.0101 | Piotr Roszatycki <dexter@cpan.org> | 2012-04-11

* Documentation tells about parsing date with year before 2000.

------------------------------------------------------------------------
0.01 | Piotr Roszatycki <dexter@cpan.org> | 2012-04-10

* First CPAN release.

------------------------------------------------------------------------


Build.PL
#!/usr/bin/perl

use 5.006;

use strict;
use warnings;

use Module::Build;

my $module = 'DateTime::Format::EMIUCP';

(my $dist = $module) =~ s/::/-/g;

my $build = Module::Build->new(
    module_name             => $module,
    license                 => 'perl',
    requires                => {
        'perl'                      => 5.006,
        'DateTime::Format::Builder' => 0,
    },
    build_requires          => {
        'Test::More'                => 0,
        'Test::Exception'           => 0,
    },
    auto_configure_requires => 0,
    create_makefile_pl      => 'traditional',
    create_license          => 1,
    recursive_test_files    => 0,
    add_to_cleanup          => [
        '*.bak', '*.gz', '*.tgz', '*.ppd', '*.ppmx', '*.deb', '*.bat',
        'Makefile', 'pm_to_blib', 'cover_db', 'Debian_CPANTS.txt',
        "$dist-*",
    ],
    meta_merge              => {
        resources                   => {
            bugtracker => "http://rt.cpan.org/NoAuth/Bugs.html?Dist=$dist",
            repository => "http://github.com/dex4er/perl-$dist",
        },
    },
);

$build->create_build_script;



lib/DateTime/Format/EMIUCP/VP.pm
package DateTime::Format::EMIUCP::VP;

=head1 NAME

DateTime::Format::EMIUCP::VP - Parse VP field for EMI-UCP protocol

=head1 SYNOPSIS

  use DateTime::Format::EMIUCP::VP;

  my $dt = DateTime::Format::EMIUCP::VP->parse_datetime('0302120655');
  print $dt->ymd; # 2012-02-03
  print $dt->hms; # 06:55:00

  $dt->set_formatter(DateTime::Format::EMIUCP::VP->new);
  print $dt; # 0302120655

=head1 DESCRIPTION

This format is a part of EMI-UCP protocol message. EMI-UCP protocol is
primarily used to connect to short message service centers (SMSCs) for mobile
telephones.

VP is a string of 10 numeric characters which represents validity period time
in ddMMyyHHmm format.

See EMI-UCP Interface 5.2 Specification for further explanations.

=for readme stop

=cut

use 5.006;

use strict;
use warnings;

our $VERSION = '0.0200';

use DateTime::Format::EMIUCP;

=head1 METHODS

=over

=item DateTime I<$dt> = $fmt->parse_datetime(Str I<$scts>)

Given a string in the pattern specified in the constructor, this method will
return a new DateTime object.

Year number below 70 means the date before year 2000.

If given a string that doesn't match the pattern, the formatter will croak.

=cut

use DateTime::Format::Builder (
    parsers => {
        parse_datetime => [
            {
                params => [qw( day month year hour minute )],
                regex  => qr/^(\d\d)(\d\d)(\d\d)(\d\d)(\d\d)$/,
                postprocess => \&_fix_year,
            },
        ]
    }
);


BEGIN { *_fix_year = \&DateTime::Format::EMIUCP::_fix_year; }


=item Str I<$scts> = $fmt->format_datetime(DateTime I<$dt>)

Given a DateTime object, this methods returns a string formatted in the
object's format.

=back

=cut

sub format_datetime {
    my ($self, $dt) = @_;
    return sprintf '%02d%02d%02d%02d%02d',
        $dt->day, $dt->month, $dt->year % 100,
        $dt->hour, $dt->minute;
};


1;

__END__

=for readme continue

=head1 PREREQUISITES

=over 2

=item *

L<DateTime::Format::Builder>

=back

=head1 SEE ALSO

L<DateTime::Format::EMIUCP>, L<DateTime>.

=head1 BUGS

If you find the bug or want to implement new features, please report it at
L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=DateTime-Format-EMIUCP>

The code repository is available at
L<http://github.com/dex4er/perl-DateTime-Format-EMIUCP>

=head1 AUTHOR

Piotr Roszatycki <dexter@cpan.org>

=head1 LICENSE

Copyright (c) 2012 Piotr Roszatycki <dexter@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as perl itself.

See L<http://dev.perl.org/licenses/artistic.html>


lib/DateTime/Format/EMIUCP/DDT.pm
package DateTime::Format::EMIUCP::DDT;

=head1 NAME

DateTime::Format::EMIUCP::DDT - Parse DDT field for EMI-UCP protocol

=head1 SYNOPSIS

  use DateTime::Format::EMIUCP::DDT;

  my $dt = DateTime::Format::EMIUCP::DDT->parse_datetime('0302120655');
  print $dt->ymd; # 2012-02-03
  print $dt->hms; # 06:55:00

  $dt->set_formatter(DateTime::Format::EMIUCP::DDT->new);
  print $dt; # 0302120655

=head1 DESCRIPTION

This format is a part of EMI-UCP protocol message. EMI-UCP protocol is
primarily used to connect to short message service centers (SMSCs) for mobile
telephones.

DDT is a string of 10 numeric characters which represents deferred delivery
time in ddMMyyHHmm format.

See EMI-UCP Interface 5.2 Specification for further explanations.

=for readme stop

=cut

use 5.006;

use strict;
use warnings;

our $VERSION = '0.0200';

use DateTime::Format::EMIUCP;

=head1 METHODS

=over

=item DateTime I<$dt> = $fmt->parse_datetime(Str I<$scts>)

Given a string in the pattern specified in the constructor, this method will
return a new DateTime object.

Year number below 70 means the date before year 2000.

If given a string that doesn't match the pattern, the formatter will croak.

=cut

use DateTime::Format::Builder (
    parsers => {
        parse_datetime => [
            {
                params => [qw( day month year hour minute )],
                regex  => qr/^(\d\d)(\d\d)(\d\d)(\d\d)(\d\d)$/,
                postprocess => \&_fix_year,
            },
        ]
    }
);


BEGIN { *_fix_year = \&DateTime::Format::EMIUCP::_fix_year; }


=item Str I<$scts> = $fmt->format_datetime(DateTime I<$dt>)

Given a DateTime object, this methods returns a string formatted in the
object's format.

=back

=cut

sub format_datetime {
    my ($self, $dt) = @_;
    return sprintf '%02d%02d%02d%02d%02d',
        $dt->day, $dt->month, $dt->year % 100,
        $dt->hour, $dt->minute;
};


1;

__END__

=for readme continue

=head1 PREREQUISITES

=over 2

=item *

L<DateTime::Format::Builder>

=back

=head1 SEE ALSO

L<DateTime::Format::EMIUCP>, L<DateTime>.

=head1 BUGS

If you find the bug or want to implement new features, please report it at
L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=DateTime-Format-EMIUCP>

The code repository is available at
L<http://github.com/dex4er/perl-DateTime-Format-EMIUCP>

=head1 AUTHOR

Piotr Roszatycki <dexter@cpan.org>

=head1 LICENSE

Copyright (c) 2012 Piotr Roszatycki <dexter@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as perl itself.

See L<http://dev.perl.org/licenses/artistic.html>


lib/DateTime/Format/EMIUCP/SCTS.pm
package DateTime::Format::EMIUCP::SCTS;

=head1 NAME

DateTime::Format::EMIUCP::SCTS - Parse SCTS field for EMI-UCP protocol

=head1 SYNOPSIS

  use DateTime::Format::EMIUCP::SCTS;

  my $dt = DateTime::Format::EMIUCP::SCTS->parse_datetime('030212065530');
  print $dt->ymd; # 2012-02-03
  print $dt->hms; # 06:55:30

  $dt->set_formatter(DateTime::Format::EMIUCP::SCTS->new);
  print $dt; # 030212065530

=head1 DESCRIPTION

This format is a part of EMI-UCP protocol message. EMI-UCP protocol is
primarily used to connect to short message service centers (SMSCs) for mobile
telephones.

SCTS is a string of 12 numeric characters which represents Service Center
time-stamp in ddMMyyHHmmss format.

See EMI-UCP Interface 5.2 Specification for further explanations.

=for readme stop

=cut

use 5.006;

use strict;
use warnings;

our $VERSION = '0.0200';

use DateTime::Format::EMIUCP;

=head1 METHODS

=over

=item DateTime I<$dt> = $fmt->parse_datetime(Str I<$scts>)

Given a string in the pattern specified in the constructor, this method will
return a new DateTime object.

Year number below 70 means the date before year 2000.

If given a string that doesn't match the pattern, the formatter will croak.

=cut

use DateTime::Format::Builder (
    parsers => {
        parse_datetime => [
            {
                params => [qw( day month year hour minute second )],
                regex  => qr/^(\d\d)(\d\d)(\d\d)(\d\d)(\d\d)(\d\d)$/,
                postprocess => \&_fix_year,
            },
        ]
    }
);


BEGIN { *_fix_year = \&DateTime::Format::EMIUCP::_fix_year; }


=item Str I<$scts> = $fmt->format_datetime(DateTime I<$dt>)

Given a DateTime object, this methods returns a string formatted in the
object's format.

=back

=cut

sub format_datetime {
    my ($self, $dt) = @_;
    return sprintf '%02d%02d%02d%02d%02d%02d',
        $dt->day, $dt->month, $dt->year % 100,
        $dt->hour, $dt->minute, $dt->second;
};


1;

__END__

=for readme continue

=head1 PREREQUISITES

=over 2

=item *

L<DateTime::Format::Builder>

=back

=head1 SEE ALSO

L<DateTime>.

=head1 BUGS

If you find the bug or want to implement new features, please report it at
L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=DateTime-Format-EMIUCP>

The code repository is available at
L<http://github.com/dex4er/perl-DateTime-Format-EMIUCP>

=head1 AUTHOR

Piotr Roszatycki <dexter@cpan.org>

=head1 LICENSE

Copyright (c) 2012 Piotr Roszatycki <dexter@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as perl itself.

See L<http://dev.perl.org/licenses/artistic.html>


lib/DateTime/Format/EMIUCP.pm
package DateTime::Format::EMIUCP;

=head1 NAME

DateTime::Format::EMIUCP - Parse time formats for EMI-UCP protocol

=head1 SYNOPSIS

  use DateTime::Format::EMIUCP;

  my $scts = DateTime::Format::EMIUCP->parse_datetime('030212065530');
  print $scts->ymd; # 2012-02-03
  print $scts->hms; # 06:55:30

  my $vp = DateTime::Format::EMIUCP->parse_datetime('0302120655');
  print $vp->ymd; # 2012-02-03
  print $vp->hms; # 06:55:00

=head1 DESCRIPTION

These formats are part of EMI-UCP protocol message. EMI-UCP protocol is
primarily used to connect to short message service centers (SMSCs) for mobile
telephones.

SCTS is a string of 12 numeric characters which represents Service Center
time-stamp in ddMMyyHHmmss format.

DDT is a string of 10 numeric characters which represents deferred delivery
time in ddMMyyHHmm format.

VP is a string of 10 numeric characters which represents validity period time
in ddMMyyHHmm format.

See EMI-UCP Interface 5.2 Specification for further explanations.

=for readme stop

=cut

use 5.006;

use strict;
use warnings;

our $VERSION = '0.0200';

=head1 METHODS

=over

=item DateTime I<$dt> = $fmt->parse_datetime(Str I<$scts>)

Given a string in the pattern specified in the constructor, this method will
return a new DateTime object.

Year number below 70 means the date before year 2000.

If given a string that doesn't match the pattern, the formatter will croak.

=back

=cut

use DateTime::Format::Builder (
    parsers => {
        parse_datetime => [
            {
                params => [qw( day month year hour minute second )],
                regex  => qr/^(\d\d)(\d\d)(\d\d)(\d\d)(\d\d)(\d\d)$/,
                postprocess => \&_fix_year,
            },
            {
                params => [qw( day month year hour minute )],
                regex  => qr/^(\d\d)(\d\d)(\d\d)(\d\d)(\d\d)$/,
                postprocess => \&_fix_year,
            },
        ]
    }
);


sub _fix_year {
    my %args = @_;
    my ($date, $p) = @args{qw( input parsed )};
    $p->{year} += $p->{year} > 69 ? 1900 : 2000;
    return 1;
};


1;

__END__

=for readme continue

=head1 PREREQUISITES

=over 2

=item *

L<DateTime::Format::Builder>

=back

=head1 SEE ALSO

L<DateTime::Format::EMIUCP::DDT>, L<DateTime::Format::EMIUCP::SCTS>,
L<DateTime::Format::EMIUCP::VP>, L<DateTime>.

=head1 BUGS

If you find the bug or want to implement new features, please report it at
L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=DateTime-Format-EMIUCP>

The code repository is available at
L<http://github.com/dex4er/perl-DateTime-Format-EMIUCP>

=head1 AUTHOR

Piotr Roszatycki <dexter@cpan.org>

=head1 LICENSE

Copyright (c) 2012 Piotr Roszatycki <dexter@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as perl itself.

See L<http://dev.perl.org/licenses/artistic.html>
